crates/core/src/sql/type_check.rs

```diff
@@ -0,0 +1,267 @@
+            // because the type is not lifted when we parse the query on `spacetimedb_vm::ops::parse`.
+            //
+            // This is a temporary solution until we have a better way to resolve the type of the field.
+            patch_type(lhs, &mut ty_lhs, &ty_rhs)?;
+            patch_type(rhs, &mut ty_rhs, &ty_lhs)?;
+
+            // If both sides are the same type, then return `Bool` to indicate a logical comparison
+            check_both(op, &ty_lhs, &ty_rhs)?;
+
+            Ok(Typed::Cmp {
+                op,
+                lhs: Box::new(ty_lhs),
+                rhs: Box::new(ty_rhs),
+                ty: AlgebraicType::Bool,
+            })
+        }
+    }
+}
+
+impl TypeCheck for QueryFragment<'_, Selection> {
+    fn type_check(&self) -> Result<(), PlanError> {
+        type_check(&QueryFragment {
+            from: self.from,
+            q: &self.q.clause,
+        })?;
+        Ok(())
+    }
+}
+
+impl TypeCheck for QueryFragment<'_, ()> {
+    fn type_check(&self) -> Result<(), PlanError> {
+        for join in &self.from.joins {
+            match join {
+                Join::Inner { rhs: _, on } => {
+                    let (table_lhs, lhs) = find_field_name(self.from, on.lhs)?;
+                    let (table_rhs, rhs) = find_field_name(self.from, on.rhs)?;
+                    let ty_lhs = resolve_type(&FieldExpr::Name(on.lhs), Some(lhs.col_type.clone()))?.unwrap();
+                    let ty_rhs = resolve_type(&FieldExpr::Name(on.rhs), Some(rhs.col_type.clone()))?.unwrap();
+
+                    if ty_lhs != ty_rhs {
+                        return Err(ErrorType::TypeMismatchJoin {
+                            lhs: Typed::Field {
+                                table: table_lhs,
+                                field: lhs.col_name.clone(),
+                                ty: Some(ty_lhs),
+                            }
+                            .to_string(),
+                            rhs: Typed::Field {
+                                table: table_rhs,
+                                field: rhs.col_name.clone(),
+                                ty: Some(ty_rhs),
+                            }
+                            .to_string(),
+                        }
+                        .into());
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl TypeCheck for SqlAst {
+    fn type_check(&self) -> Result<(), PlanError> {
+        match self {
+            SqlAst::Select {
+                from,
+                project: _,
+                selection,
+            } => {
+                QueryFragment { from, q: &() }.type_check()?;
+                if let Some(selection) = selection {
+                    QueryFragment { from, q: selection }.type_check()?;
+                }
+            }
+
+            _ => {
+                // TODO: Other options deferred for the new query engine
```

Comment:
The discussion with Joshua was to keep this as a temporal solution because the proposal for `SQL` APIs touches this in ways that will invalidate the changes.